### PR TITLE
[CO-427] Foundation for external wallets, WIP!

### PR DESCRIPTION
**Pull request by denisshevchenko**
## Description

This PR provides a foundation for work with external wallets. It includes:

1. `UserPublic` and `UserSecret` corrections.
2. Wallet user key instead of encrypted secret key.
3. Create/restore external wallets.
4. Delete external wallets.

Please note that endpoint `checkExternalWallet` was removed for now. This endpoint was described by Daedalus team [in old CHW task](https://iohk.myjetbrains.com/youtrack/issue/CHW-52#comment=93-22582). But since we don't have `checkWallet` endpoint, I decided to remove it for now. We will add it in the future if Daedalus team really needs it (and they will confirm it).

Handlers in `API/WIP/Handlers.hs` are used now. Handlers in `API/WIP/LegacyHandlers.hs` was removed (replaced by `error`-messages), because their code used the old `wallet` (and we want to get rid of this dependency).

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CO-427

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: wallet-new/src/Cardano/Wallet/Kernel/Restore.hs [hunk 1]
```diff
@@ -218,9 +293,13 @@ continueRestoration pw root cur tgt = do
             -- TODO (@mn): raise an error, trying to continue
             -- restoration of an unknown wallet
             return ()
-        Just esk -> do
-            let prefilter = mkPrefilter pw wId esk
-                wkey      = (wId, keyToWalletDecrCredentials (KeyForRegular esk))
+        Just walletUserKey -> do
+            let prefilter = mkPrefilter pw wId walletUserKey
```
Comment (paweljakubas) [inline wallet-new/src/Cardano/Wallet/Kernel/Restore.hs]: comment like above

Comment (paweljakubas) [inline wallet-new/src/Cardano/Wallet/Kernel/Restore.hs]: comment like above

File: wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs [hunk 2]
```diff
@@ -78,6 +83,19 @@ createWallet wallet newWalletRequest = liftIO $ do
                                       (HD.WalletName newwalName)
       return (mkRoot newwalName newwalAssuranceLevel now root)
 
+    createExternal :: V1.NewExternalWallet
+                   -> Timestamp
+                   -> IO (Either CreateWalletError V1.Wallet)
+    createExternal V1.NewExternalWallet{..} now = runExceptT $ do
```
Comment (paweljakubas) [inline wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs]: I like very much that `runExceptT` is used

File: wallet/test/Test/Pos/Wallet/Arbitrary/Web/ClientTypes.hs [hunk 3]
```diff
@@ -17,7 +17,7 @@ import           Pos.Wallet.Web.ClientTypes.Types (CHash (..), CId (..),
 import           Pos.Wallet.Web.State (WAddressMeta (..))
 
 import           Test.Pos.Core.Arbitrary ()
-import           Universum
+import           Universum ()
```
Comment (paweljakubas) [inline wallet/test/Test/Pos/Wallet/Arbitrary/Web/ClientTypes.hs]: I do not think it is good change - `(<$>)`, `pure` and `(.)` used below are not in scope then.

File: wallet-new/src/Cardano/Wallet/Kernel/DB/AcidState.hs [hunk 4]
```diff
@@ -529,6 +531,47 @@ createHdWallet newRoot defaultHdAccountId defaultHdAddress utxoByAccount =
...
+--
+-- NOTE: We allow an initial set of accounts with associated addresses and
+-- balances /ONLY/ for testing purpose. Normally this should be empty; see
+-- 'createHdWallet'/'createWalletHdRnd' in "Cardano.Wallet.Kernel.Wallets".
+--
+-- INVARIANT: Creating a new wallet always come with a fresh HdAccount and
+-- a fresh 'HdAddress' attached to it, so we have to pass these two extra
+-- piece of into to the update function. We do @not@ build these inside the
+-- update function because derivation requires an 'EncryptedSecretKey' and
+-- definitely we do not want it to show up in our acid-state logs.
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/Kernel/DB/AcidState.hs]: This comment is wrong and copy-pasted from above I believe. We do not create any address in the HdWallet. 
This invariant was primarily a sugar for Daedalus and other clients, so I understand why we do not preserve it for the HdWallet; 
@denisshevchenko, any extra inputs maybe?

File: wallet-new/src/Cardano/Wallet/API/WIP.hs [hunk 5]
```diff
@@ -12,10 +12,6 @@ import           Cardano.Wallet.API.V1.Types
 type API = Tags '["WIP"] :>
     (
          "external-wallets"
-                   :> Capture "rootPK" PublicKeyAsBase58
-                   :> Summary "Check if this external wallet is presented in the node."
-                   :> PostCreated '[ValidJSON] (WalletResponse WalletAndTxHistory)
-    :<|> "external-wallets"
                    :> Summary "Creates a new or restores an existing external wallet (mobile client or hardware wallet)."
                    :> ReqBody '[ValidJSON] (New ExternalWallet)
                    :> PostCreated '[ValidJSON] (WalletResponse Wallet)
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/WIP.hs]: Shouldn't that response type be `WalletResponse ExternalWallet` or am I missing something here :thinking: ?

File: wallet-new/src/Cardano/Wallet/Kernel/DB/AcidState.hs [hunk 6]
```diff
@@ -529,6 +531,47 @@ createHdWallet newRoot defaultHdAccountId defaultHdAddress utxoByAccount =
...
+--
+createExternalHdWallet :: HdRoot
+                       -> HdAccountId
+                       -- ^ The default HdAccountId to go with this HdRoot. This
+                       -- function will take responsibility of creating the associated
+                       -- 'HdAccount'.
+                       -> Map HdAccountId (Utxo, [AddrWithId])
+                       -> Update DB (Either HD.CreateHdRootError ())
+createExternalHdWallet newRoot defaultHdAccountId utxoByAccount =
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/Kernel/DB/AcidState.hs]: Do external wallets share the same account id representation than the current wallets?

File: wallet-new/src/Cardano/Wallet/Kernel/Pending.hs [hunk 7]
```diff
@@ -108,24 +108,26 @@ newTx ActiveWallet{..} accountId tx partialMeta upd = do
...
+        map f $ filterOurs wKey txOutAddress txOut
+      where
+        f (txOut',addressId) = (initHdAddress addressId (txOutAddress txOut'), txOutValue txOut')
+        wKey = case walletUserKey of
+                   RegularWalletKey esk -> (wid, keyToWalletDecrCredentials $ KeyForRegular esk)
+                   ExternalWalletKey pk -> (wid, keyToWalletDecrCredentials $ KeyForExternal pk)
+
+    submitTx :: IO ()
+    submitTx = modifyMVar_ (walletPassive ^. walletSubmission) $
+                return . addPending accountId (Pending.singleton tx)
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/Kernel/Pending.hs]: What's happening here :/ ? May we avoid re-indentation of the whole block to make the changes obvious (EncryptedSecretKey -> WalletUserKey I believe)

File: wallet-new/src/Cardano/Wallet/API/V1/ReifyWalletError.hs [hunk 8]
```diff
@@ -179,18 +185,25 @@ updateWalletPasswordError e = case e of
...
         (Kernel.UpdateWalletPasswordUnknownHdRoot (HD.UnknownHdRoot _rootId)) ->
-                V1.WalletNotFound
+            V1.WalletNotFound
 
         ex@(Kernel.UpdateWalletPasswordKeystoreChangedInTheMeantime _rootId) ->
-                V1.UnknownError (sformat build ex)
+            V1.UnknownError (sformat build ex)
+
+        ex@Kernel.UpdateWalletPasswordUnableForExternalWallet ->
+            V1.UnknownError (sformat build ex)
```
[the remark this answers is not in the document]
Comment (denisshevchenko) [reply wallet-new/src/Cardano/Wallet/API/V1/ReifyWalletError.hs]: Yes, why not? We're happy with `UnknownError` for `UpdateWalletPasswordKeystoreChangedInTheMeantime`, why not here?

File: wallet-new/src/Cardano/Wallet/API/WIP.hs (hunk 5 above)
Comment (denisshevchenko) [reply wallet-new/src/Cardano/Wallet/API/WIP.hs]: No, it should return `Wallet`, for unification. We _can_ think of external wallet as of regular one, but without spending password and with root PK only. :-)

File: wallet-new/src/Cardano/Wallet/Kernel/DB/AcidState.hs [hunk 9]
```diff
@@ -577,6 +620,47 @@ restoreHdWallet newRoot defaultHdAccountId defaultHdAddress ctx utxoByAccount =
...
+        }
+
+    insertDefault :: Map HdAccountId (Utxo, Utxo, [AddrWithId])
+                  -> Map HdAccountId (Utxo, Utxo, [AddrWithId])
+    insertDefault m = case Map.lookup defaultHdAccountId m of
+        Just _ -> m
+        Nothing ->
+            -- Completely new account, without addresses (yet).
+            Map.insert defaultHdAccountId (mempty, mempty, []) m
+
```
[the remark this answers is not in the document]
Comment (denisshevchenko) [reply wallet-new/src/Cardano/Wallet/Kernel/DB/AcidState.hs]: Fixed.

File: wallet-new/src/Cardano/Wallet/Kernel/Keystore.hs [hunk 10]
```diff
@@ -161,68 +185,101 @@ bracketTestKeystore withKeystore =
...
-insert :: WalletId -> EncryptedSecretKey -> Keystore -> IO ()
-insert _walletId esk ks = modifyKeystore_ ks (insertKey esk)
+insert :: WalletUserKey -> Keystore -> IO ()
+insert (RegularWalletKey rootSK) ks  = modifyKeystoreSecret_ ks (insertSecretKey rootSK)
+insert (ExternalWalletKey rootPK) ks = modifyKeystorePublic_ ks (insertPublicKey rootPK)
 
 -- | Insert a new 'EncryptedSecretKey' directly inside the 'UserSecret'.
-insertKey :: EncryptedSecretKey -> UserSecret -> UserSecret
-insertKey esk us =
+insertPublicKey :: PublicKey -> UserPublic -> UserPublic
```
[the remark this answers is not in the document]
Comment (denisshevchenko) [reply wallet-new/src/Cardano/Wallet/Kernel/Keystore.hs]: Fixed.

File: wallet-new/src/Cardano/Wallet/Kernel/Keystore.hs [hunk 11]
```diff
@@ -266,27 +323,51 @@ compareAndReplace walletId predicateOnOldKey newKey ks =
...
+    let mbEsk = lookupSecretKey us walletId
+        erase = Data.List.deleteBy (\k1 k2 -> hash k1 == hash k2)
     in maybe us (\esk -> us & over usKeys (erase esk)) mbEsk
+
+-- | Try to delete a key directly inside the 'UserPublic'.
+tryToDeletePublicKey :: WalletId -> UserPublic -> UserPublic
+tryToDeletePublicKey walletId up =
+    let mbPK = lookupPublicKey up walletId
+        erase = Data.List.deleteBy (\k1 k2 -> hash k1 == hash k2)
+    in maybe up (\pk -> up & over upKeys (erase pk)) mbPK
```
[the remark this answers is not in the document]
Comment (denisshevchenko) [reply wallet-new/src/Cardano/Wallet/Kernel/Keystore.hs]: Removed duplication.

File: wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs [hunk 12]
```diff
@@ -586,12 +589,13 @@ instance Arbitrary SignTransactionError where
 
 mkSigner :: NetworkMagic
          -> PassPhrase
-         -> Maybe EncryptedSecretKey
+         -> Maybe Keystore.WalletUserKey
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs]: That's not the right place to make this separation. `mkSigner` should accept an `EncryptedSecretKey`. The caller has to deal with whether it has a key or not.

File: wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs [hunk 13]
```diff
@@ -570,6 +570,7 @@ data SignTransactionError =
     SignTransactionMissingKey Address
   | SignTransactionErrorUnknownAddress Address
   | SignTransactionErrorNotOwned Address
+  | SignTransactionUnableForExternalWallet
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs]: Not sure we'll need this. It'd be better if we enforce that only "classic" wallets can be signed.

File: wallet-new/src/Cardano/Wallet/Kernel/Wallets.hs [hunk 14]
```diff
@@ -88,6 +92,8 @@ data UpdateWalletPasswordError =
       -- ^ When trying to update the password inside the keystore, the
       -- previous 'PassPhrase' didn't match or it was deleted, which means
       -- this operation is not valid anymore.
+    | UpdateWalletPasswordUnableForExternalWallet
+      -- External wallets don't have spending password by definition.
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/Kernel/Wallets.hs]: Okay for now, but we'll move forward with defining another type for external wallet which won't have passwords. Then, it won't be necessary to have such error.

**Comment by denisshevchenko** [comment]
Closed in favor of https://github.com/input-output-hk/cardano-sl/pull/3808.

**Pull request closed without merging**